tfworker/backends/s3.py

```diff
@@ -302,3 +321,171 @@ def filter_keys(paginator, bucket_name, prefix="/", delimiter="/", start_after="
                     yield content["Key"]
         except TypeError:
             pass
+
+
+class S3Handler(BaseHandler):
+    """The S3Handler class is a handler for the s3 backend"""
+
+    actions = ["plan", "apply"]
+    required_vars = []
+    _is_ready = False
+
+    def __init__(self, authenticator):
+        try:
+            self.execution_functions = {
+                "plan": {
+                    "check": self._check_plan,
+                    "post": self._post_plan,
+                },
+                "apply": {
+                    "pre": self._pre_apply,
+                },
+            }
+
+            self._authenticator = authenticator
+            self._s3_client = self._authenticator.backend_session.client("s3")
+
+        except Exception as e:
+            raise HandlerError(f"Error initializing S3Handler: {e}")
+
+    def is_ready(self):
+        if not self._is_ready:
+            filename = str(uuid4().hex[:6].upper())
+            if self._s3_client.list_objects(
+                Bucket=self._authenticator.bucket,
+                Prefix=f"{self._authenticator.prefix}/{filename}",
+            ).get("Contents"):
+                raise HandlerError(
+                    f"Error initializing S3Handler, remote file already exists: {filename}"
+                )
+            try:
+                self._s3_client.upload_file(
+                    "/dev/null",
```

Comment:
Is this essentially a 'touch' of the file in s3?

Reply:
Yup, it seemed like it would be better to know before hand if there's permissions issues than finding out after the plan is already done and attempted to be stashed.